ApplySchema tries to reload schema on slaves.

enisoc opened the pull request:
@michael-berlin 

After applying a schema change to the master of each shard, `vtctl ApplySchema` will now tell each slave tablet to reload schema once it receives the DDLs through MySQL replication. In the happy path, this makes the slave tablets appear to reload the new schema immediately. Previously, you had to either wait for the periodic reload as configured by the vttablet `-queryserver-config-schema-reload-time` flag, or send `vtctl ReloadSchema` to each tablet manually.

If any slaves take longer than `-wait_slave_timeout` (as specified on the `ApplySchema` command) to confirm the schema reload, we will print warnings to indicate that the operator should retry the ReloadSchema command on that tablet.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/1806)
<!-- Reviewable:end -->

sougou commented on go/vt/tabletserver/schema_info.go:
```diff
@@ -229,11 +230,28 @@ func (si *SchemaInfo) Close() {
 	defer logError(si.queryServiceStats)
-	ctx := context.Background()
+
+	if si.IsClosed() {
```
This is racy because si can get closed immediately after the check. The better way will be to use the startRequest pattern at the tabletserver level.

sougou commented on go/vt/tabletserver/schema_info.go:
```diff
@@ -291,29 +312,24 @@ func (si *SchemaInfo) mysqlTime(ctx context.Context) int64 {
 
 // CreateOrUpdateTable must be called if a DDL was applied to that table.
-func (si *SchemaInfo) CreateOrUpdateTable(ctx context.Context, tableName string) {
+func (si *SchemaInfo) CreateOrUpdateTable(ctx context.Context, tableName string) error {
```
I think this function and DropTable should acquire the reloadSchema lock.

sougou replied:
I'm thinking we can get rid of both functions and just call ReloadSchema instead.

michael-berlin commented:
Reviewed 39 of 58 files at r1.
Review status: 39 of 58 files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[go/vt/mysqlctl/mysql_daemon.go, line 330 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPTPNwIH2BcI0cto20:-KKPTPNwIH2BcI0cto21:-1904659701) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/mysql_daemon.go#L330)):*
> ```Go
> 
> // WaitMasterPos is part of the MysqlDaemon interface
> func (fmd *FakeMysqlDaemon) WaitMasterPos(ctx context.Context, pos replication.Position) error {
> ```

optional: If a parameter is unused, you can set its name to "_". Same comment applies to the other changes in this file.

---

*[go/vt/mysqlctl/query.go, line 78 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPWlQiB6aVQjq6kRD0:-KKPWlQiB6aVQjq6kRD1:-966246114) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L78)):*
> ```Go
> // executeFetchContext calls Execute() on the given connection,
> // while respecting Context deadline and cancellation.
> func (mysqld *Mysqld) executeFetchContext(ctx context.Context, conn dbconnpool.PoolConnection, query string, maxrows int, wantfields bool) (*sqltypes.Result, error) {
> ```

As discussed offline: Ctrl + F does not find any users of this function at this point.

---

*[go/vt/mysqlctl/query.go, line 90 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPVfsYhNIRaFIexBZv:-KKPVfsYhNIRaFIexBZw:1375000588) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L90)):*
> ```Go
> 	}
> 
> 	// Execute the query.
> ```

Mention here that you call it asynchronously to be able to check its completion and the context Done channel at the same time?

---

*[go/vt/mysqlctl/query.go, line 111 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPVX0swbAZkcgArGa-:-KKPVX0t2IfCuOI0C_lJ:1301144882) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L111)):*
> ```Go
> 		}
> 
> 		// Try to kill the connection.
> ```

Please add a reason why you want to kill the connection here.

---

*[go/vt/mysqlctl/query.go, line 158 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPWbjX0AxvxMNISgRu:-KKPWbjX0AxvxMNISgRv:-1385548274) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L158)):*
> ```Go
> 	}
> 
> 	_, err := killConn.ExecuteFetch(fmt.Sprintf("kill %d", connID), 10000, false)
> ```

Any particular reason why you have 10k for maxrows? Is it worth a comment?

---

*[go/vt/vtctld/api.go, line 225 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPkEMBAe_ZZVWUShWO:-KKPkEMChuL9YgwmF8NF:410965403) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/vtctld/api.go#L225)):*
> ```Go
> 	http.HandleFunc(apiPrefix+"schema/apply", func(w http.ResponseWriter, r *http.Request) {
> 		req := struct {
> 			Keyspace, SQL       string
> ```

optional: I'm a fan of one field per line. But I'm fine with the current version :)

---

*[go/vt/wrangler/schema.go, line 64 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPlNQUgEg634SwOMOg:-KKPlNQVuD4vOSBHZyNb:867283078) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/wrangler/schema.go#L64)):*
> ```Go
> 		}
> 		// We got a partial result. Do what we can, but warn that some may be missed.
> 		wr.logger.Warningf("Failed to fetch all tablets for %v/%v. Some slave tablets may not have schema reloaded (use vtctl ReloadSchema to fix individual tablets)", keyspace, shard)
> ```

Is it worth to have this special case? Does it happen that often?

Should we not just error here instead?

---

*[go/vt/wrangler/schema.go, line 67 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPlzpdhCB3qjckVOdt:-KKPlzpdhCB3qjckVOdu:984224231) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/wrangler/schema.go#L67)):*
> ```Go
> 	}
> 
> 	wg := &sync.WaitGroup{}
> ```

Why get the pointer here instead of the instance?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1806)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
I've reviewed everything now.

LGTM after the comments are resolved.

---

Reviewed 19 of 58 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks broke.

---

*[go/vt/schemamanager/tablet_executor.go, line 200 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQP_dcgYTUfIBpg0U6:-KKQP_dcgYTUfIBpg0U7:-1822729929) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L200)):*
> ```Go
> 	defer func() { execResult.TotalTimeSpent = time.Since(startTime) }()
> 
> 	// lock the keyspace
> ```

nit: "Lock the keyspace."

---

*[go/vt/schemamanager/tablet_executor.go, line 233 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQQ44f_LLSxaV9q3vB:-KKQQ44f_LLSxaV9q3vC:405178337) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L233)):*
> ```Go
> 
> func (exec *TabletExecutor) executeOnAllTablets(ctx context.Context, execResult *ExecuteResult, sql string) {
> 	wg := &sync.WaitGroup{}
> ```

No need for storing the pointer instead of the object?

---

*[go/vt/schemamanager/tablet_executor.go, line 263 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQQOBIf9Mpq6Cycxst:-KKQQOBIf9Mpq6Cycxsu:-400165279) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L263)):*
> ```Go
> 	// execute the schema change via replication. This is best-effort, meaning
> 	// we still return overall success if the timeout expires.
> 	wg = &sync.WaitGroup{}
> ```

You could reuse the waitGroup from above?

Alternatively, I would prefer to not reuse variable names and give this a different name.

---

*[go/vt/tabletmanager/rpc_schema.go, line 35 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQNq-21AfEXGhQ45cj:-KKQNq-21AfEXGhQ45ck:-956391160) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletmanager/rpc_schema.go#L35)):*
> ```Go
> 	}
> 
> 	if waitPosition != "" {
> ```

optional: You could define a constant for this. The code would read more naturally then. For example, name the constant "currentPosition":

```go
if waitPosition != currentPosition {
```

---

*[go/vt/tabletserver/schema_info.go, line 273 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQHrLB7weKz5Oo8hBk:-KKQHrLB7weKz5Oo8hBl:-95858463) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletserver/schema_info.go#L273)):*
> ```Go
> 	// The following section requires us to hold mu.
> 	errs := &concurrency.AllErrorRecorder{}
> 	func() {
> ```

The extra func() is probably not necessary?

I assume the intent here was to scope the "defer". But based on the current code, it would get executed on the end of the function anyway.

---

*[test/local_example.sh, line 66 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQOSLUeO2rTyzsa4F2:-KKQOSLUeO2rTyzsa4F3:-1843385567) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/test/local_example.sh#L66)):*
> ```Shell
> # This is not necessary, but it helps make this test more representative of
> # what a human would do. It simulates the case where a periodic health check
> # occurs before the user gets around to running the next command.
> ```

It's not clear to me which implications a healthcheck us: What's the reason you want to simulate this? Maybe add this to the comment here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1806#-:-KKQSDik8j_aioOn-OU_:721063571)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Review status: 0 of 58 files reviewed at latest revision, 16 unresolved discussions.

---

*[go/vt/mysqlctl/mysql_daemon.go, line 330 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPTPNwIH2BcI0cto20:-KKUbP352k_PRMIlzWRd:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/mysql_daemon.go#L330)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> optional: If a parameter is unused, you can set its name to "_". Same comment applies to the other changes in this file.
</details>
Done.

---

*[go/vt/mysqlctl/query.go, line 78 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPWlQiB6aVQjq6kRD0:-KKUaB5EVe_0yOOzGhFJ:-529017311) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L78)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> As discussed offline: Ctrl + F does not find any users of this function at this point.
</details>
Oops! I forgot to change the call sites after writing this function... Fixed.

---

*[go/vt/mysqlctl/query.go, line 90 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPVfsYhNIRaFIexBZv:-KKUamixUwfuwhKxP8kR:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L90)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Mention here that you call it asynchronously to be able to check its completion and the context Done channel at the same time?
</details>
Done.

---

*[go/vt/mysqlctl/query.go, line 111 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPVX0swbAZkcgArGa-:-KKUcAcYR65AIgLhwYS9:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L111)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Please add a reason why you want to kill the connection here.
</details>
Done.

---

*[go/vt/mysqlctl/query.go, line 158 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPWbjX0AxvxMNISgRu:-KKUcE0u96XLIM2gJ3u1:710983385) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L158)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Any particular reason why you have 10k for maxrows? Is it worth a comment?
</details>
No reason that's specific to this invocation. The 10k value is cargo-culted all over the codebase as a value that means "more than zero, but less than infinity". If you want to fix it, we should do that separately and fix all occurrences.

---

*[go/vt/schemamanager/tablet_executor.go, line 200 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQP_dcgYTUfIBpg0U6:-KKUd3CehLkcC3fMdGYn:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L200)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> nit: "Lock the keyspace."
</details>
Done.

---

*[go/vt/schemamanager/tablet_executor.go, line 233 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQQ44f_LLSxaV9q3vB:-KKUdFT-0BMcn95ncU08:390029567) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L233)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> No need for storing the pointer instead of the object?
</details>
Oops. I made this change to make it easier to do the right thing when passing `wg` to child functions. Then I decided passing `wg` to child functions at all is a bad idea. Forgot to revert this.

---

*[go/vt/schemamanager/tablet_executor.go, line 263 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQQOBIf9Mpq6Cycxst:-KKUe8EyRl_NaqVCSZpI:1710785287) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L263)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> You could reuse the waitGroup from above?
> 
> Alternatively, I would prefer to not reuse variable names and give this a different name.
</details>
At the time, I checked WaitGroup docs and didn't see any guarantees that it's safe to reuse them. I just checked again and I did find that guarantee in the docs for `Add()`.

---

*[go/vt/tabletmanager/rpc_schema.go, line 35 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQNq-21AfEXGhQ45cj:-KKUgRh3cHzUvsrKjW9A:-1683176738) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletmanager/rpc_schema.go#L35)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> optional: You could define a constant for this. The code would read more naturally then. For example, name the constant "currentPosition":
> 
> ```go
> if waitPosition != currentPosition {
> ```
</details>
I'd rather think of it as "wait position is empty; therefore don't wait" rather than "wait position is a magic constant that means current position".

---

*[go/vt/tabletserver/schema_info.go, line 245 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKMAPLnZEp63FBfP35z-r1-245:-KKV0jydwfL8mB4oV1G1:-1845974846) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletserver/schema_info.go#L245)):*
<details><summary><i>Previously, sougou wrote…</i></summary>
> This is racy because si can get closed immediately after the check. The better way will be to use the startRequest pattern at the tabletserver level.
</details>
I looked at `startRequest` and it seems more complex than what I need here. I believe I've addressed the races by using an "action mutex" concept, similar to tabletmanager.

---

*[go/vt/tabletserver/schema_info.go, line 273 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQHrLB7weKz5Oo8hBk:-KKV1JhvtBFJ73bm8iVl:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletserver/schema_info.go#L273)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> The extra func() is probably not necessary?
> 
> I assume the intent here was to scope the "defer". But based on the current code, it would get executed on the end of the function anyway.
</details>
Done.

---

*[go/vt/tabletserver/schema_info.go, line 321 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKMAPLnZEp63FBfP35z-r1-321:-KKPUUSP_K2W78BECvQ9:1697125730) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletserver/schema_info.go#L321)):*
<details><summary><i>Previously, sougou wrote…</i></summary>
> I'm thinking we can get rid of both functions and just call ReloadSchema instead.
</details>
I added the actionMutex to DropTable and CreateOrUpdateTable, but kept the direct calls from QueryExector. I used the `func *Locked()` convention to avoid double locking via Reload().

If you think it's better to just call Reload() from QueryExecutor, I think that's a separate consideration orthogonal to locking.

---

*[go/vt/vtctld/api.go, line 225 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPkEMBAe_ZZVWUShWO:-KKV1XO2l59LOMB2puIF:-189672619) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/vtctld/api.go#L225)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> optional: I'm a fan of one field per line. But I'm fine with the current version :)
</details>
I would agree for a named struct, especially if it's part of a public API. But for a one-off anonymous struct, I appreciate brevity.

---

*[go/vt/wrangler/schema.go, line 64 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPlNQUgEg634SwOMOg:-KKV1nVelGJehtZzxcx9:-1994033702) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/wrangler/schema.go#L64)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Is it worth to have this special case? Does it happen that often?
> 
> Should we not just error here instead?
</details>
GetTabletMapForShard first lists the tablets and then does a GetTablet on each one.  It's not uncommon for one of the GetTablet calls to fail out of hundreds. In that case, we don't want to abort the whole thing, especially since that one failure may be non-transient which would make retrying the ReloadSchemaShard fail every time. Since ReloadSchemaShard is itself guaranteeing only best-effort reloads, we should continue with whatever we have.

---

*[go/vt/wrangler/schema.go, line 67 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPlzpdhCB3qjckVOdt:-KKV2_Gwe6_V8GjKRBci:1296404362) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/wrangler/schema.go#L67)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Why get the pointer here instead of the instance?
</details>
No reason. Fixed.

---

*[test/local_example.sh, line 66 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQOSLUeO2rTyzsa4F2:-KKV3zhkkMuw6lSptqRW:-499146009) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/test/local_example.sh#L66)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> It's not clear to me which implications a healthcheck us: What's the reason you want to simulate this? Maybe add this to the comment here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1806)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
Reviewed 13 of 13 files at r3, 7 of 7 files at r4, 15 of 17 files at r5, 10 of 10 files at r6, 1 of 1 files at r7, 6 of 8 files at r8, 2 of 2 files at r9, 1 of 1 files at r10, 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[go/vt/mysqlctl/query.go, line 158 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKPWbjX0AxvxMNISgRu:-KKVjh8sk4_G-8rFm1vD:2053930743) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/mysqlctl/query.go#L158)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> No reason that's specific to this invocation. The 10k value is cargo-culted all over the codebase as a value that means "more than zero, but less than infinity". If you want to fix it, we should do that separately and fix all occurrences.
</details>
No worries. I already thought this was the case. I'm fine with keeping the tradition :)

---

*[go/vt/schemamanager/tablet_executor.go, line 263 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKQQOBIf9Mpq6Cycxst:-KKVk-DQC-DuEj9IYfBt:-528254886) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/schemamanager/tablet_executor.go#L263)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> At the time, I checked WaitGroup docs and didn't see any guarantees that it's safe to reuse them. I just checked again and I did find that guarantee in the docs for `Add()`.
</details>
FYI: We're also reusing WaitGroups in the resharding code e.g. split_clone.go.

---

*[go/vt/tabletserver/schema_info.go, line 245 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1806#-KKMAPLnZEp63FBfP35z-r1-245:-KKVkJ-TrvdNQuuKsWdQ:-1517891314) ([raw file](https://github.com/youtube/vitess/blob/85430143d250fa37a82f34b9252a4a51979946e0/go/vt/tabletserver/schema_info.go#L245)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> I looked at `startRequest` and it seems more complex than what I need here. I believe I've addressed the races by using an "action mutex" concept, similar to tabletmanager.
</details>
I discussed with with Sugu on Wednesday and he was fine with using a mutex in all calls as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1806)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
LGTM
